go/processor/floria/calls.go

```
@@ -16,62 +16,70 @@ import (
+		Depth:                 r.depth - 1, // depth is already incremented
+		Gas:                   parameters.Gas,
+		Recipient:             parameters.Recipient,
+		Sender:                parameters.Sender,
+		Input:                 parameters.Input,
+		Value:                 parameters.Value,
+		CodeHash:              &codeHash,
+		Code:                  code,
 	}
 
-	snapshot := context.CreateSnapshot()
-	if err := transferValue(transaction, context); err != nil {
-		context.RestoreSnapshot(snapshot)
-		return tosca.Result{}, nil
+	snapshot := r.CreateSnapshot()
+	if err := transferValue(r, parameters.Value, parameters.Sender, parameters.Recipient); err != nil {
+		r.RestoreSnapshot(snapshot)
+		return tosca.CallResult{}, nil
 	}
 
-	result, err := interpreter.Run(params)
+	result, err := r.interpreter.Run(params)
 	if err != nil || !result.Success {
-		context.RestoreSnapshot(snapshot)
+		r.RestoreSnapshot(snapshot)
 	}
 
-	return result, err
+	return tosca.CallResult{
+		Output:    result.Output,
+		GasLeft:   result.GasLeft,
+		GasRefund: result.GasRefund,
+		Success:   result.Success,
+	}, err
 }
 
-func transferValue(transaction tosca.Transaction, context tosca.TransactionContext) error {
-	senderBalance := context.GetBalance(transaction.Sender)
-	transferValue := transaction.Value
+func transferValue(
+	context tosca.TransactionContext,
+	value tosca.Value,
+	sender tosca.Address,
+	recipient tosca.Address,
+) error {
+	senderBalance := context.GetBalance(sender)
+	transferValue := value
 	if senderBalance.Cmp(transferValue) < 0 {
 		return fmt.Errorf("insufficient balance: %v < %v", senderBalance, transferValue)
 	}
 
 	senderBalance = tosca.Sub(senderBalance, transferValue)
-	context.SetBalance(transaction.Sender, senderBalance)
+	context.SetBalance(sender, senderBalance)
 
-	receiverBalance := context.GetBalance(*transaction.Recipient)
+	receiverBalance := context.GetBalance(recipient)
 	receiverBalance = tosca.Add(receiverBalance, transferValue)
```

If we want to play it super-safe we would have to check that adding the transferred value is not leading to an overflow at the receiver. In such a case, we would also let the transfer fail.

---

Right, this has been missed in the last PR were this function was introduced. I restructured the whole function and added more unit tests.